sshd-core/src/main/java/org/apache/sshd/common/session/helpers/KeyExchangeMessageHandler.java

```diff
@@ -0,0 +1,485 @@
+    }
+
+    /**
+     * Writes a packet. If a key exchange is ongoing, only low-level messages are written directly; all other messages
+     * are queued and will be written once {@link #flushQueue(DefaultKeyExchangeFuture)} is called when the key exchange
+     * is done. Packets written while there are still pending packets to be flushed will either be queued, too, or the
+     * calling thread will be blocked with the given timeout until all packets have been flushed. Whether a write will
+     * be blocked is determined by {@link #isBlockAllowed(int)}.
+     * <p>
+     * If {@code timeout <= 0} or {@code unit == null}, a time-out of "forever" is assumed. Note that a timeout applies
+     * only if the calling thread is blocked.
+     * </p>
+     *
+     * @param  buffer      packet to write
+     * @param  timeout     number of {@link TimeUnit}s to wait at most if the calling thread is blocked
+     * @param  unit        {@link TimeUnit} of {@code timeout}
+     * @return             an {@link IoWriteFuture} that will be fulfilled once the packet has indeed been written.
+     * @throws IOException if an error occurs
+     */
+    public IoWriteFuture writePacket(Buffer buffer, long timeout, TimeUnit unit) throws IOException {
+        // While exchanging key, queue high level packets.
+        byte[] bufData = buffer.array();
+        int cmd = bufData[buffer.rpos()] & 0xFF;
+        boolean enqueued = false;
+        boolean isLowLevelMessage = cmd <= SshConstants.SSH_MSG_KEX_LAST && cmd != SshConstants.SSH_MSG_SERVICE_REQUEST
+                && cmd != SshConstants.SSH_MSG_SERVICE_ACCEPT;
+        try {
+            if (isLowLevelMessage) {
+                // Low-level messages can always be sent.
+                return session.doWritePacket(buffer);
+            }
+            IoWriteFuture future = writeOrEnqueue(cmd, buffer, timeout, unit);
+            enqueued = future instanceof PendingWriteFuture;
+            return future;
+        } finally {
+            session.resetIdleTimeout();
+            if (!enqueued) {
+                try {
+                    session.checkRekey();
+                } catch (GeneralSecurityException e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("writePacket({}) failed ({}) to check re-key: {}", session, e.getClass().getSimpleName(),
+                                e.getMessage(), e);
+                    }
+                    throw ValidateUtils.initializeExceptionCause(new ProtocolException(
+                            "Failed (" + e.getClass().getSimpleName() + ")" + " to check re-key necessity: " + e.getMessage()),
+                            e);
+                } catch (Exception e) {
+                    ExceptionUtils.rethrowAsIoException(e);
+                }
+            }
+        }
+    }
+
+    private IoWriteFuture writeOrEnqueue(int cmd, Buffer buffer, long timeout, TimeUnit unit) throws IOException {
```

Comment:
Let's make this *protected* in the spirit of letting our users override whatever they like.

Reply:
Done.